packages/journey-manager/src/index.ts

```diff
@@ -376,12 +326,73 @@ export const run = (props: RunProps): RunOutput => {
     return [];
   };
 
+  /**
+   * UI management
+   */
+
+  // TODO: type this more accurately
+  let uiElement: any;
+
+  let teardownUiElement: (() => void) | null = null;
+
+  if (props.ui != null) {
+    uiElement = document.createElement("journey-manager");
+    uiElement.style.zIndex = 1000;
+    uiElement.config = props.ui;
+    uiElement.client = client;
+    const handleAction = (ev: any): void => {
+      const action = ev.detail?.action;
+      if (action == null) {
+        return;
+      }
+      if (action === "escalate" && props.ui?.escalationStep != null) {
+        client.sendStep(props.ui.escalationStep).catch((err) => {
+          // eslint-disable-next-line no-console
+          console.warn(err);
+        });
+        return;
+      }
+      if (action === "end" && props.ui?.endStep != null) {
+        client.sendStep(props.ui.endStep).catch((err) => {
+          // eslint-disable-next-line no-console
+          console.warn(err);
+        });
+        return;
+      }
+      if (action === "previous") {
+        const lastTriggeredStep = triggeredSteps[triggeredSteps.length - 1];
+        if (lastTriggeredStep != null) {
+          client.sendStep(lastTriggeredStep).catch((err) => {
+            // eslint-disable-next-line no-console
+            console.warn(err);
+          });
+        }
+      }
+    };
+    let highlightInterval: NodeJS.Timeout | null = null;
+    if (props.ui.highlights === true) {
+      highlightInterval = setInterval(() => {
+        const highlightElements = findActiveTriggers("click").flatMap(
+          (activeTrigger) => activeTrigger.elements,
+        );
+        uiElement.highlightElements = highlightElements;
+      }, 500);
+    }
```

Comment:
Why not mutation observer? In fact, we could probably even build a faster `matchesActiveTriggers` that would fit neatly into mutation observer callbacks.

Reply:
Here it is: https://github.com/nlxai/sdk/issues/103